### PR TITLE
ROC-8031 paper response admission event changes

### PR DESCRIPTION
**Pull request by tribhuwan-hmcts**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/ROC-8031

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandler.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+import static uk.gov.hmcts.cmc.claimstore.services.notifications.NotificationReferenceBuilder.ResponseSubmitted.referenceForClaimant;
+import static uk.gov.hmcts.cmc.claimstore.services.notifications.NotificationReferenceBuilder.ResponseSubmitted.referenceForDefendant;
+import static uk.gov.hmcts.cmc.domain.utils.LocalDateTimeFactory.UTC_ZONE;
+
+@Service
+public class PaperResponseAdmissionCallbackHandler extends CallbackHandler {
+
+    private static final List<CaseEvent> EVENTS = List.of(PAPER_RESPONSE_ADMISSION);
+    private static final List<Role> ROLES = List.of(CASEWORKER);
+    private static final String FORMNAME = "OCON9x";
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandler.java:50]: `Form name` should be camel cased
```suggestion
    private static final String FORM_NAME = "OCON9x";
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java [hunk 2]
```diff
@@ -199,4 +199,22 @@ private CCDAddress getDefendantAddress(CCDRespondent respondent) {
             ? respondent.getPartyDetail().getPrimaryAddress()
             : respondent.getClaimantProvidedDetail().getPrimaryAddress();
     }
+
+    public DocAssemblyTemplateBody paperResponseAdmissionLetter(CCDCase ccdCase, String caseworkerName) {
+        LocalDate currentDate = LocalDate.now(clock.withZone(UTC_ZONE));
+        String partyName = ccdCase.getRespondents().get(0)
+            .getValue().getPartyName() != null
+            ? ccdCase.getRespondents().get(0).getValue().getPartyName() :
+            ccdCase.getRespondents().get(0).getValue().getClaimantProvidedPartyName();
```
Comment (daveoh) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/legaladvisor/DocAssemblyTemplateBodyMapper.java:208]: A bit easier to read if you set the defendant to a variable
```suggestion
        CCDRespondent defendant =  ccdCase.getRespondents().get(0).getValue();
        String partyName = defendant.getPartyName() != null
            ? defendant.getPartyName() 
            :  defendant.getClaimantProvidedPartyName();
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+import static uk.gov.hmcts.cmc.ccd.domain.CCDClaimDocumentType.GENERAL_LETTER;
+import static uk.gov.hmcts.cmc.ccd.domain.CCDPartyType.INDIVIDUAL;
+import static uk.gov.hmcts.cmc.ccd.domain.CaseEvent.PAPER_RESPONSE_ADMISSION;
+import static uk.gov.hmcts.cmc.ccd.domain.defendant.CCDResponseType.FULL_ADMISSION;
+import static uk.gov.hmcts.cmc.ccd.domain.defendant.CCDResponseType.PART_ADMISSION;
+import static uk.gov.hmcts.cmc.claimstore.services.ccd.Role.CASEWORKER;
+
+@ExtendWith(MockitoExtension.class)
+@DisplayName("Paper Response Admission handler")
+public class PaperResponseAdmissionCallbackHandlerTest {
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java:64]: Access should be default for Junit 5 tests
```suggestion
class PaperResponseAdmissionCallbackHandlerTest {
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    @Mock
+    private Clock clock;
+    @Mock
+    private UserService userService;
+    @Mock
+    private DocAssemblyTemplateBodyMapper docAssemblyTemplateBodyMapper;
+    private UserDetails userDetails;
+
+    @BeforeEach
+    public void setUp() {
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java:92]: Access should be default for Junit 5 tests
```suggestion
   void setUp() {
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        assertEquals("CMC-scanned-OCON9x-part-admission.pdf",
+            ccdDataArgumentCaptor.getValue().getScannedDocuments().get(0).getValue().getFileName());
+        assertNull(ccdDataArgumentCaptor.getValue().getPaperAdmissionType());
+        assertEquals(LocalDateTime.of(2020, Month.JUNE, 17, 11, 30, 57),
+            ccdDataArgumentCaptor.getValue().getRespondents().get(0).getValue().getResponseSubmittedOn());
+        assertEquals(1, ccdDataArgumentCaptor.getValue().getScannedDocuments().size());
+    }
+
+    @Test
+    public void shouldNotChangeFileNameForOtherForm() {
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java:184]: Access should be default for Junit 5 tests
```suggestion
   void shouldNotChangeFileNameForOtherForm() {
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+        when(caseDetailsConverter.extractCCDCase(any(CaseDetails.class))).thenReturn(ccdCase);
+
+        Assertions.assertThrows(IllegalStateException.class,
+            () -> handler.handle(callbackParams));
+
+    }
+
+    @Test
+    public void notifyClaimant() {
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java:196]: Access should be default for Junit 5 tests
```suggestion
   void notifyClaimant() {
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+        handler.handle(callbackParams);
+
+        verify(defendantResponseNotificationService, times(1)).notifyClaimant(
+            any(Claim.class),
+            any(String.class)
+        );
+    }
+
+    @Test
+    public void notifyDefendant() {
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java:216]: Access should be default for Junit 5 tests
```suggestion
   void notifyDefendant() {
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+        when(docAssemblyTemplateBodyMapper.paperResponseAdmissionLetter(any(CCDCase.class), any(String.class)))
+            .thenReturn(DocAssemblyTemplateBody.builder().build());
+        when(caseDetailsConverter.extractCCDCase(any(CaseDetails.class))).thenReturn(ccdCase);
+
+        Assertions.assertThrows(RuntimeException.class,
+            () -> handler.handle(callbackParams));
+    }
+
+    @Test
+    public void shouldHaveCorrectCaseworkerRole() {
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java:322]: Access should be default for Junit 5 tests
```suggestion
   void shouldHaveCorrectCaseworkerRole() {
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+        assertThat(handler.getSupportedRoles()).containsOnly(CASEWORKER);
+    }
+
+    @Test
+    public void shouldHaveCorrectCaseworkerEvent() {
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java:327]: Access should be default for Junit 5 tests
```suggestion
   void shouldHaveCorrectCaseworkerEvent() {
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+            .renderTemplate(any(CCDCase.class), anyString(), anyString(), any(DocAssemblyTemplateBody.class)))
+            .thenReturn(docAssemblyResponse);
+        when(docAssemblyTemplateBodyMapper.paperResponseAdmissionLetter(any(CCDCase.class), any(String.class)))
+            .thenReturn(DocAssemblyTemplateBody.builder().build());
+        when(docAssemblyResponse.getRenditionOutputLocation()).thenReturn(DOC_URL);
+        when(documentManagementService.getDocumentMetaData(anyString(), anyString())).thenReturn(getLinks());
+        when(clock.instant()).thenReturn(LocalDate.parse("2020-06-22").atStartOfDay().toInstant(ZoneOffset.UTC));
+        when(clock.getZone()).thenReturn(ZoneOffset.UTC);
+        when(clock.withZone(LocalDateTimeFactory.UTC_ZONE)).thenReturn(clock);
+        when(caseDetailsConverter.extractCCDCase(any(CaseDetails.class))).thenReturn(ccdCase);
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java:136]: Looks like these are repeated in this and `shouldChangeFileNameForOCON9xFormForPartAdmission`. You should move this into a nested sub class and put this in a @BeforeEach method. Give me a shout if you want help.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java (hunk 7 above)
Comment (daveoh) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java:216]: Also test name should be 
```suggestion
   void shouldNotifyDefendantByEmail() {
```

File: src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+
+        handler.handle(callbackParams);
+        verify(userService, times(1)).getUserDetails(anyString());
```
Comment (daveoh) [inline src/test/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandlerTest.java:260]: I dont think we need to test this. Its not a requirement of the story.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandler.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+
+    private void sendClaimantEmail(Claim claim) {
+        defendantResponseNotificationService.notifyClaimant(
+            claim,
+            referenceForClaimant(claim.getReferenceNumber())
+        );
+    }
+
+    private CCDCase sendDefendantLetter(CallbackParams callbackParams, CCDCase updatedCCDCase) {
+        String authorisation = callbackParams.getParams().get(CallbackParams.Params.BEARER_TOKEN).toString();
```
Comment (dharmendrak) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandler.java:139]: This method doesn't send letter for bulk print

Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/caseworker/PaperResponseAdmissionCallbackHandler.java:139]: we could have used the GeneralLetterService to generate and send the letter

**Pull request merged**